lib/ddtrace/contrib/httparty/request_patch.rb

```diff
@@ -0,0 +1,137 @@
+require 'ddtrace/ext/net'
+require 'ddtrace/ext/distributed'
+require 'ddtrace/propagation/http_propagator'
+require 'ddtrace/contrib/rest_client/ext'
+
+module Datadog
+  module Contrib
+    module HTTParty
+      # HTTParty RequestPatch
+      module RequestPatch
+        def self.included(base)
+          base.send(:prepend, InstanceMethods)
+        end
+
+        # InstanceMethods - implementing instrumentation
+        module InstanceMethods
+          def uri
+            # uri is a method not a property in underlying class
+            # cache it so that it is only called once per call to 'perform', and then delete it
+            @dd_uri ||= super
+          end
+
+          def perform(&block)
+            return super unless tracer.enabled
+
+            datadog_trace_request(http_method::METHOD, uri) do
+              super
+            end
+          ensure
+            # delete cached uri
+            remove_instance_variable('@dd_uri') if instance_variable_defined?('@dd_uri')
+          end
+
+          private
+
+          def datadog_start_trace(method, uri)
+            tracer.trace(Ext::SPAN_REQUEST,
+                         service: service_name,
+                         span_type: Datadog::Ext::HTTP::TYPE_OUTBOUND).tap do |span|
+              # add tags to span
+              span.resource = method.to_s.upcase
+              span.set_tag(Datadog::Ext::HTTP::URL, uri.path)
+              span.set_tag(Datadog::Ext::HTTP::METHOD, method.to_s.upcase)
+              span.set_tag(Datadog::Ext::NET::TARGET_HOST, uri.host)
+              span.set_tag(Datadog::Ext::NET::TARGET_PORT, uri.port)
+
+              # Set analytics sample rate
+              Contrib::Analytics.set_sample_rate(span, analytics_sample_rate) if analytics_enabled?
+
+              if distributed_tracing?
+                # add span context to http headers
+                headers = options[:headers].respond_to?(:to_hash) ? options[:headers].to_hash : {}
+                Datadog::HTTPPropagator.inject!(span.context, headers)
+                options[:headers] = headers
+              end
+            end
+          end
+
+          def datadog_finish_trace(span, response)
+            return if span.nil? || !response.is_a?(::HTTParty::Response)
+
+            span.set_tag(Datadog::Ext::HTTP::STATUS_CODE, response.code)
+          end
+
+          def datadog_trace_request(method, uri)
```

Comment:
This method seems very duplicative of `Tracer#trace`: is there a simpler way of doing this? Haven't yet wrapped my head around the instrumentation strategy. If you could offer some brief explanation of how HTTParty performs requests and our instrumentation strategy here, that would be helpful.

Reply:
updated